### PR TITLE
Idr0021

### PR DESCRIPTION
**Pull request by jburel**
Discussed with @pwalczysko 
This script combines the 3 initially required for the IDR dataset
The images are analysed using the map annotation to determine the channel to analyse, a summary is then created and saved as a CSV file (one row per image)
If the server is not readonly, 
the ROIs are saved, the CVS file is uploaded and attached to the project and an OMERO table is created from the CVS
The CSV can then be used in the R notebook to plot the result
or the user can open the CSV

### CODE REVIEW

**Comment by jburel** [comment]
@pwalczysko @will-moore: groovy version of the full script is now done.
As discussed I will split the script into small script for the training

**Comment by will-moore** [comment]
Locally, I created a Project containing just the first 2 Datasets to make script a bit quicker...

Jython produced a csv with one line extended with extra values and then the OMERO.table creation failed (tried to put 'CENT2' into a Double column).

Tried Groovy script on the same data and after processing all the images it failed with:
screen shot 2018-08-14 at 16 12 08
which doesn't give much info.

Maybe I'll try on eel server instead of localhost...

**Comment by will-moore** [comment]
CSV file that was attached to Project from Jython script here: (renamed to txt to allow github upload)
[_var_folders_b5__dnzdy750_s53w_21t6lvf61cn9pqx_T_idr0021_merged_results_jfojmZ.txt](https://github.com/ome/training-scripts/files/2287512/_var_folders_b5__dnzdy750_s53w_21t6lvf61cn9pqx_T_idr0021_merged_results_jfojmZ.txt)

**Comment by jburel** [comment]
@will-moore problem should now be fixed. I have removed the usage of the tmp CSV files

**Comment by will-moore** [comment]
With a Project containing just the first 2 Datasets I'm seeing a problem with Table creation:

screen shot 2018-08-17 at 11 34 52

I'll try with fewer images (to save time) and try print out the table data...

**Comment by will-moore** [comment]
Hmmm - this seems to be working for me now.
Maybe I had an old tmp csv open / or not removed.
I'll try with the full set of images....

**Comment by will-moore** [comment]
OK, so it finally completed for all 413 images and created an OMERO.table.
However, OMERO.parade is having trouble because the numpy histogram fails with some of the Columns of data, in particular ```Total_Area```.
Histograms are calculated for all columns when we try to load a Table filter. When iterating through the columns, we are OK until we get to ```Total_Area``` but then it hangs (development server on my machine) until the devserver is killed.
```
Col Total Area 413
Killed: 9
```
When I try to "Add Table data..." this works for various columns but not for ```Total Area```.
This fails when trying to create a Histogram. Looking at the data I get back when I comment-out the histogram creation I see some numbers that are very different, which might explain the problem:
```
"76031": 0.8659388422966003,
"76032": 0.7614020705223083,
"76033": 0.39630353450775146,
"76034": 1.1545851230621338,
"76035": 0.9985600709915161,
"76036": 56252376,
"76037": 51523196,
"76038": 47665180,
"76039": 56127924,
"76040": 77160336,
"76041": 79649384,
"76042": 177842144,
"76043": 59239228,
"76044": 0.7025356888771057,
"76045": 0.8969874978065491,
"76046": 0.6445137858390808,
"76047": 0.7182172536849976,
"76048": 0.352836012840271,
```

**Comment by will-moore** [comment]
The reason these 8 images have a much larger area (in microns) is because the pixel size is wrong. See the first 8 images from the NEDD1 Dataset at http://idr.openmicroscopy.org/webclient/?show=image-1885071  ```Pixels Size (XYZ) (µm):  352.78 x 352.78 x -```.

Either we need to ignore pixels size and simply use pixels or use the same pixel size for all images.
All others seem to be ```0.04 x 0.04 x 0.13```.

**Comment by jburel** [comment]
The best option here is to exclude those 8 images from the analysis
since we will indicate that the same script can be used against IDR data
The workaround can then be removed when we fix the metadata in IDR

**Comment by jburel** [comment]
@will-moore I have now removed the .tif files

**Comment by jburel** [comment]
@will-moore I have removed the images and parade now behave
I still think we have issue with parade itself with certain type of values.

File: practical/groovy/idr0021.groovy [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * ------------------------------------------------------------------------------
+ */
+
+/*
+ * This Groovy script uses ImageJ to analyse particles, and saved the results
+ * locally in a CSV file. The file is then uploaded and attached to the specified
```
Comment (pwalczysko) [inline practical/groovy/idr0021.groovy]: There is no mention here about the script adding a table as well as CSV. I can see in the body of the script that the table is added as well (whcih is good, I think this just needs to be clarified in the header here)

File: practical/groovy/idr0021.groovy [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+ * project as a file annotation.
+ * Use this script in the Scripting Dialog of Fiji (File > New > Script).
+ * Select Groovy as language in the Scripting Dialog.
+ * Error handling is omitted to ease the reading of the script but
+ * this should be added
+ * if used in production to make sure the services are closed
+ * Information can be found at
+ * https://docs.openmicroscopy.org/latest/omero/developers/Java.html
```
Comment (pwalczysko) [inline practical/groovy/idr0021.groovy]: ``.../omero5/...`` is in https://github.com/ome/training-scripts/pull/31/files#diff-7b76ac0d9cb22c3981f59e92b5a80790R34 - why is it different here ? (``.../omero/...``)

**Comment by pwalczysko** [comment]
@jburel Now I am getting

```
Started New_.groovy at Mon Aug 20 14:39:16 BST 2018
groovy.lang.MissingMethodException: No signature of method: java.lang.String.endWith() is applicable for argument types: (java.lang.String) values: [.tif]
Possible solutions: endsWith(java.lang.String), length(), with(groovy.lang.Closure), next(), indexOf(java.lang.String), split(java.lang.String)
	at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.unwrap(ScriptBytecodeAdapter.java:58)
	at org.codehaus.groovy.runtime.callsite.PojoMetaClassSite.call(PojoMetaClassSite.java:49)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at Script3$_run_closure1$_closure4.doCall(Script3.groovy:436)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:294)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1027)
	at groovy.lang.Closure.call(Closure.java:414)
	at groovy.lang.Closure.call(Closure.java:430)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2030)
...
```

**Comment by pwalczysko** [comment]
Now the script ran to the end successfully. It processed 405 images.

But, the names of the Datasets in the Dataset column in some cases do not match the names of the Datasets in OMERO. For example I have ``PCNT-N20`` name as a dataset name in the CSV - but such Dataset name does not exist in idr0021. Also, the graph is not matching, I am completely missing ``CPAP`` for example and the other points are hard to understand. Plotted Dataset names against the Bounding Box.
screen shot 2018-08-20 at 17 59 27

I explored the image ID 9725 which is in the CPAP dataset in OMERO. But in the CSV produced by this script it is in the PCNT-N20 Dataset, which does not exist in OMERO... Thinking that maybe CPAP name somehow metamorphosed to be PCNT-N20, I compared the graph with the publication and that did not match either.

**Comment by pwalczysko** [comment]
The jython script ends up with 

```
Started New_.groovy at Mon Aug 20 16:28:36 BST 2018
Started New_.py at Mon Aug 20 18:22:29 BST 2018
Traceback (most recent call last):
  File "New_.py", line 417, in <module>
  File "New_.py", line 194, in save_row
	at ij.measure.ResultsTable.getStringValue(ResultsTable.java:417)
	at ij.measure.ResultsTable.getStringValue(ResultsTable.java:406)
	at sun.reflect.GeneratedMethodAccessor35.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)

java.lang.IllegalArgumentException: java.lang.IllegalArgumentException: Column not defined: 27

	at org.python.core.Py.JavaError(Py.java:546)
	at org.python.core.Py.JavaError(Py.java:537)
	at org.python.core.PyReflectedFunction.__call__(PyReflectedFunction.java:188)
	at org.python.core.PyReflectedFunction.__call__(PyReflectedFunction.java:204)
	at org.python.core.PyObject.__call__(PyObject.java:496)
	at org.python.core.PyObject.__call__(PyObject.java:500)
	at org.python.core.PyMethod.__call__(PyMethod.java:156)
	at org.python.pycode._pyx0.save_row$7(New_.py:218)
	at org.python.pycode._pyx0.call_function(New_.py)
	at org.python.core.PyTableCode.call(PyTableCode.java:167)
	at org.python.core.PyBaseCode.call(PyBaseCode.java:307)
	at org.python.core.PyFunction.function___call__(PyFunction.java:471)
	at org.python.core.PyFunction.__call__(PyFunction.java:466)
	at org.python.core.PyFunction.__call__(PyFunction.java:456)
	at org.python.pycode._pyx0.f$0(New_.py:440)
	at org.python.pycode._pyx0.call_function(New_.py)
	at org.python.core.PyTableCode.call(PyTableCode.java:167)
	at org.python.core.PyCode.call(PyCode.java:18)
	at org.python.core.Py.runCode(Py.java:1386)
	at org.python.core.__builtin__.eval(__builtin__.java:497)
	at org.python.core.__builtin__.eval(__builtin__.java:501)
	at org.python.util.PythonInterpreter.eval(PythonInterpreter.java:259)
	at org.python.jsr223.PyScriptEngine.eval(PyScriptEngine.java:40)
	at org.python.jsr223.PyScriptEngine.eval(PyScriptEngine.java:31)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.scijava.script.ScriptModule.run(ScriptModule.java:159)
	at org.scijava.module.ModuleRunner.run(ModuleRunner.java:167)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:126)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:65)
	at org.scijava.thread.DefaultThreadService$3.call(DefaultThreadService.java:237)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalArgumentException: Column not defined: 27
	at ij.measure.ResultsTable.getStringValue(ResultsTable.java:417)
	at ij.measure.ResultsTable.getStringValue(ResultsTable.java:406)
	at sun.reflect.GeneratedMethodAccessor35.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.python.core.PyReflectedFunction.__call__(PyReflectedFunction.java:186)
	... 31 more
```

**Comment by pwalczysko** [comment]
If this was tested on eel, then possibly on http://web-dev-merge.openmicroscopy.org/webclient/?show=project-4501 ? On that project, note that Parade does not work really. Tried on Bounding box, which gives just three possible values. Then tried mean area, and the whole central pane vanished. Just like you reported with Trackmate.

**Comment by jburel** [comment]
The error is a bit strange

**Comment by jburel** [comment]
I ran the jython script this evening and no error
``
Found index: 1
creating summary results for image ID 118514
1
/var/folders/k2/g_p4x1_x0f7_wsfc7sldtr500000gr/T/idr0021_merged_results_aV6agb.csv
processing done
``

**Comment by will-moore** [comment]
Running the current script against outreach gives me a csv that has no ```CPAP``` in the Dataset column.
Also, it would be nice to have a Channel column to show which channels was analysed for ROI size

**Comment by will-moore** [comment]
To summarise, script is working fine. Could benefit from having a Channel column.
Also, the csv has a random (temp) name, which could be a bit neater. 
e.g. idr0021_merged_results.csv instead of idr0021_merged_results_hTyFKh.csv

**Comment by pwalczysko** [comment]
```
Started New_.groovy at Thu Aug 23 15:16:08 BST 2018
java.lang.IllegalArgumentException: Column not defined: 27
	at ij.measure.ResultsTable.getStringValue(ResultsTable.java:417)
	at ij.measure.ResultsTable.getStringValue(ResultsTable.java:406)
	at ij.measure.ResultsTable$getStringValue$1.call(Unknown Source)
	at Script1.save_row(Script1.groovy:221)
	at sun.reflect.GeneratedMethodAccessor32.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:384)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1027)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.callCurrent(PogoMetaClassSite.java:69)
	at Script1$_run_closure1$_closure4.doCall(Script1.groovy:478)
	at sun.reflect.GeneratedMethodAccessor20.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:294)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1027)
	at groovy.lang.Closure.call(Closure.java:414)
	at groovy.lang.Closure.call(Closure.java:430)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2030)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2015)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2056)
	at org.codehaus.groovy.runtime.dgm$162.invoke(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite$PojoMetaMethodSiteNoUnwrapNoCoerce.invoke(PojoMetaMethodSite.java:274)
	at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.call(PojoMetaMethodSite.java:56)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at Script1$_run_closure1.doCall(Script1.groovy:441)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:294)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1027)
	at groovy.lang.Closure.call(Closure.java:414)
	at groovy.lang.Closure.call(Closure.java:430)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2030)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2015)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2068)
	at org.codehaus.groovy.runtime.dgm$164.invoke(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite$PojoMetaMethodSiteNoUnwrapNoCoerce.invoke(PojoMetaMethodSite.java:274)
	at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.call(PojoMetaMethodSite.java:56)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at Script1.run(Script1.groovy:436)
	at org.scijava.plugins.scripting.groovy.GroovyScriptEngine.eval(GroovyScriptEngine.java:303)
	at org.scijava.plugins.scripting.groovy.GroovyScriptEngine.eval(GroovyScriptEngine.java:122)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.scijava.script.ScriptModule.run(ScriptModule.java:159)
	at org.scijava.module.ModuleRunner.run(ModuleRunner.java:167)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:126)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:65)
	at org.scijava.thread.DefaultThreadService$3.call(DefaultThreadService.java:237)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by pwalczysko** [comment]
The way to repeat the problem https://github.com/ome/training-scripts/pull/30#issuecomment-415443374

  - start Fiji, open scripts dialog etc. as described in the header of the PR
  - but have a running Firefox previously opened (Fiji is opened on top of that FF window)
  - paste the script, edit the necessary fields and click Run
  - observe that images are being open, ROIs drawn, and measurements harvested for several images, no problems observed
  - click onto the underlying Firefox window to continue your work in Firefox
  - go back to Fiji
  - observe that the script crashed in the moment you navigated away from Fiji to Firefox
  - happens for both groovy and jython script

**Comment by pwalczysko** [comment]
So I have just re-run the full jython script on idr0021 outreach trainer-1. There are no other CSV attached, so yoiu might see yourself.

There is no CPAP in my Dataset column...

I really do not know. For sure, one of you guys would have to please re-run the script yourselves, to end up with the right CSV @jburel @will-moore

**Comment by pwalczysko** [comment]
I have re-run the jython script on eel. Now, we have also here http://web-dev-merge.openmicroscopy.org/webclient/?show=project-4501 just csv which has no CPAP in the Dataset column - looks similar to the one on outreach.

**Comment by jburel** [comment]
@pwalczysko I think I have finally found the source of the problem

**Comment by pwalczysko** [comment]
The groovy and jython scripts both produce the same CSV, which makes sense.

The follow-up R testing shows that the tables are also consistent and match the CSVs. 

All  good here.
Ready to merge FMPOV

**Pull request merged**